Update pagination docs (#163)
MicrosoftDocs/OData-docs: Odata-docs/client/v7/pagination.md

---
title: "Pagination"
description: "This tutorial describes how to deal with server-driven and client-driven pagination."

author: mumbi-o
ms.author: mowambug
ms.date: 7/1/2019
ms.topic: article
 
---
# Pagination
**Applies To**: [!INCLUDE[appliesto-odataclient](../../includes/appliesto-odataclient-v7.md)]
Loading large datasets can be slow. Services often rely on pagination to load the data incrementally to improve the response times and the user experience. Paging can be server-driven or client-driven.

## Server-driven paging

> In Server-driven paging, the server returns the first page of results. If total number of results is greater than the page size, the server returns the first page along with a nextlink that can be used to fetch the next page of results.

Reviewer: ```suggestion
In Server-driven paging, the server returns the first page of results. If total number of results is greater than the page size, the server returns the first page along with a `nextLink` that can be used to fetch the next page of results.
```

Author: #Resolved

> The OData Client deals with server-driven paging with the help of [DataServiceQueryContinuation](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation) and [DataServiceQueryContinuation&lt;T&gt;](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation-1). These classes contain the nextLink of the partial set of items.

Reviewer: ```suggestion
The OData Client deals with server-driven paging with the help of [DataServiceQueryContinuation](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation) and [DataServiceQueryContinuation&lt;T&gt;](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation-1). These classes contain the `nextLink` of the partial set of items.
```

Author: #Resolved

### Top Level Pagination

Example:

``` csharp
DefaultContainer context = new DefaultContainer(new Uri("https://services.odata.org/v4/TripPinServiceRW/"));

// DataServiceQueryContinuation<T> contains the next link
DataServiceQueryContinuation<Person> nextLink = null;

// Get the first page
QueryOperationResponse<Person> response = await context.People.ExecuteAsync() as QueryOperationResponse<Person>;
int pageCount = 0;

do
{
    Console.WriteLine($"Page {++pageCount}");
    if (nextLink != null)
    {
        response = await context.ExecuteAsync<Person>(nextLink) as QueryOperationResponse<Person>;
    }

    // You must enumerate the response before calling GetContinuation below.
    foreach (Person person in response)
    {
        Console.WriteLine($"\tPerson Name: {person.FirstName}");
    }

}
// Loop if there is a next link
while ((nextLink = response.GetContinuation()) != null);
```

### Nested Pagination
There are instances where we need to load pages of entities as well as pages of their related entities.

> The example below returns related Trips for each Person entity from the data service. It uses a do...while loop to paginate through Person entities and a nested while loop to paginate through the related Trips.

Reviewer: ```suggestion
The example below returns related `Trips` for each `Person` entity from the data service. It uses a do...while loop to paginate through `Person` entities and a nested while loop to paginate through the related `Trips`.
```

Author: #Resolved

---
title: "Pagination"
description: "This tutorial describes how to deal with server-driven and client-driven pagination."

author: mumbi-o
ms.author: mowambug
ms.date: 7/1/2019
ms.topic: article
 
---
# Pagination
**Applies To**: [!INCLUDE[appliesto-odataclient](../../includes/appliesto-odataclient-v7.md)]

> Loading large datasets can be slow. Services often rely on pagination to load the data incrementally to improve the response times and the user experience. Paging can be server-driven or client-driven:

Reviewer 2: ```suggestion
Loading large datasets can be slow. Services often rely on pagination to load the data incrementally to improve the response times and the user experience. Paging can be server-driven or client-driven.
```

Author: #Resolved

## Server-driven paging
In Server-driven paging, the server returns the first page of results. If total number of results is greater than the page size, the server returns the first page along with a nextlink that can be used to fetch the next page of results.

> The OData Client deals with server-driven paging with the help of [DataServiceQueryContinuation](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation) and [DataServiceQueryContinuation&lt;T&gt;](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation-1). They are classes that contain the next link of the partial set of items.

Reviewer 2: ```suggestion
The OData Client deals with server-driven paging with the help of [DataServiceQueryContinuation](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation) and [DataServiceQueryContinuation&lt;T&gt;](/dotnet/api/microsoft.odata.client.dataservicequerycontinuation-1). These classes contain the next link of the partial set of items.
```

Author: #Resolved

### Top Level Pagination

Example:

``` csharp
DefaultContainer context = new DefaultContainer(new Uri("https://services.odata.org/v4/TripPinServiceRW/"));

// DataServiceQueryContinuation<T> contains the next link
DataServiceQueryContinuation<Person> nextLink = null;

// Get the first page
QueryOperationResponse<Person> response = await context.People.ExecuteAsync() as QueryOperationResponse<Person>;
int pageCount = 0;

do
{
    pageCount++;

>     Console.WriteLine($"Page {pageCount}");

Reviewer 2: ```suggestion
    Console.WriteLine($"Page {++pageCount}");
```
and delete line above this one

        }

        // Enumerate the Person(s) in the response.
        foreach (Person person in response)
        {
            var innerPageCount = 0;
            Console.WriteLine($"\tPerson Name: {person.FirstName}");

            // Get the next link for the collection of related Trips.
            DataServiceQueryContinuation tripsNextLink =
                response.GetContinuation(person.Trips);

            if (person.Trips.Count > 0)
            {
                Console.WriteLine($"\t\tTrips Page {++innerPageCount}:");
            }
            foreach (Trip trip in person.Trips)
            {
                // Print out the trips in the first page.

>                 Console.WriteLine($"\t\t\tTripID: {trip.TripId} - Name: ${trip.Name}");

Reviewer 2: ```suggestion
                Console.WriteLine($"\t\t\tTripID: {trip.TripId} - Name: {trip.Name}");
```

Author: #Resolved

            }

            while (tripsNextLink != null)
            {
                // Load the next page of Trips.
                var tripsResponse = await context.LoadPropertyAsync(person, "Trips", tripsNextLink);
                tripsNextLink = tripsResponse.GetContinuation();

                if (tripsResponse.Count > 0)
                {
                    Console.WriteLine($"\t\tTrips Page {++innerPageCount}:");
                }
                foreach (Trip trip in tripsResponse)
                {
                    // Print out the trips.

>                     Console.WriteLine($"\t\t\tTripID: {trip.TripId} - Name: ${trip.Name}");

Reviewer 2: ```suggestion
                    Console.WriteLine($"\t\t\tTripID: {trip.TripId} - Name: {trip.Name}");
```

Author: #Resolved

               .............................................
               .............................................
```

## Client-driven paging

> In client-driven paging, we request the server to return the specified number of results. There is no nextLink that is returned.

Reviewer: ```suggestion
In client-driven paging, we request the server to return the specified number of results. There is no `nextLink` that is returned.
```

Author: #Resolved

               .............................................
               .............................................
```

## Client-driven paging

> In client-driven paging, we request the server to return the specified number of results. There is no next link that is returned.

Reviewer 2: ```suggestion
In client-driven paging, we request the server to return the specified number of results. There is no nextLink that is returned.
```

Author: #Resolved

> The OData Client deals with client-driven paging using `$skip` and `$top` in the query options.

Reviewer 2: ```suggestion
The OData Client deals with client-driven paging using `$skip` and `$top` query options.
```

Author: #Resolved
